pimcore/models/Tool/CustomReport/Config.php

```diff
@@ -150,7 +150,23 @@ public static function getReportsList()
     public static function getAdapter($configuration, $fullConfig = null)
     {
         $type = $configuration->type ? ucfirst($configuration->type) : 'Sql';
-        $adapter = "\\Pimcore\\Model\\Tool\\CustomReport\\Adapter\\{$type}";
+        if ($type === 'Custom') {
+            if (@\class_exists($configuration->className, true)) { // can need autoloader, can send warning and break json response so we use @
+                $adapter = $configuration->className;
+                $adpaterInstance = new $adapter($configuration, $fullConfig);
+                // we make some check for security reason (because className can be choosen at the GUI level)
+                if (get_parent_class($adpaterInstance) !== 'Pimcore\\Model\\Tool\\CustomReport\\Adapter\\AbstractAdapter') {
```

Comment:
Yes i totally agree it is a lot more secure this way. +1 for me

But in the end the best impementation would be like newsletter so both work the same way. But it need to rewrite a little the current reports adapter (sql, analytics) by implementing interface instead abstracted class. And recode js part too to allow report namespace to have adpater added (for listing adapters added in the GUI). This will be the best solution i think as developper will see consistency coding pratic between newletter/report for adding custom newsletter/report mecanism.